### DISCUSSION TITLE
Any easy way to attach additional span processor?

### DISCUSSION THREAD

**Discussion by patsonluk** [open-telemetry/opentelemetry-java#3033] [Q&A]
We want to experiment with some extra functionality(profiling) that it's critical to trigger:
1. on the same thread as the instrumentation
2. immediately (synchronously) when a span is created and ended

It appears that having our own implementation of `SpanProcessor` would be the most ideal candidate, however, with the existing design, the easiest approach for configuration will be something similar to https://github.com/open-telemetry/opentelemetry-java/blob/main/QUICKSTART.md#tracing-sdk-configuration. 

While it's not too terrible, it would have been much nicer if it can be built with some configuration option via [auto config](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/OpenTelemetrySdkAutoConfiguration.java). It does seem to be a bit tricky to configure any additional span processor for auto config as the parameters right now are pretty simple (which is good), having an extra span processor config option could be messy (what exporter/sampler to attach to that)

I am wondering if there are any thoughts or suggestions for our usage? Are there something else other than `SpanProcessor` that we can look at?

**Comment by jkwatson** [comment]
Hi @patsonluk . This is possible today, by using the autoconfigure SPI interface: `SdkTracerProviderConfigurer`. By implementing this interface, you can add your own `SpanProcessor` at that time. If you combine this with setting your exporter option to `none`, then you'll have complete control over how your `TracerProvider` is built. If you simply want an additional SpanProcessor, then you can just have your `SdkTracerProviderConfigurer` implementation add a SpanProcessor and let the rest of the autoconfigure system do its thing.

 Will that work for now?

**Reply by patsonluk** [reply]
Ah I have totally missed it! Thank you so much @jkwatson ! It works perfectly ;)

**Reply by patsonluk** [reply]
Oh and a quick question, is there any way to access the `ConfigProperties config` in `TracerProviderConfiguration.configureTracerProvider(Resource resource, ConfigProperties config)` from the `SdkTracerProviderConfigurer` ? I could see that `configurer.configure(tracerProviderBuilder);` is invoked on the `SdkTracerProviderConfigurer` but seems like the `config` is not available? 

We want to do some initialization to our version of `SdkTracerProviderConfigurer` base on the `config`. Not a huge deal right now as we can simply call `System.getProperties`. Or perhaps I overlooked and it's actually available? 😅

**Reply by jkwatson** [reply]
Yes, I have been wanting to update the SPI to have the `config` passed in, but just haven't gotten around to it. PRs welcome!  ;)
